### PR TITLE
[Merged by Bors] - [Merged by Bors] - feat: more expr traversal functions

### PR DESCRIPTION
**Pull request by EdAyers**
- split Mathlib/Lean/Expr into multiple files.
- implement a version of replaceRecM that correctly instantiates free vars.
- some additonal Expr functions

These are prereqs of #234  and #229

### CODE REVIEW

File: Mathlib/Lean/Expr/Basic.lean [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+def natLit! : Expr → Nat
+  | lit (Literal.natVal v) _ => v
+  | _                        => panic! "nat literal expected"
+
+/-- Returns a `NameSet` of all constants in an expression starting with a certain prefix. -/
+def listNamesWithPrefix (pre : Name) (e : Expr) : NameSet :=
+  e.foldConsts ∅ fun n l => if n.getPrefix == pre then l.insert n else l
+
+def modifyAppArgM [Functor M] [Pure M] (modifier : Expr → M Expr) : Expr → M Expr
+  | e@(app f a _) => e.updateApp! f <$> modifier a
```
Comment (gebner) [inline Mathlib/Lean/Expr/Basic.lean:121]:
```suggestion
  | app f a _ => mkApp f <$> modifier a
```

File: Mathlib/Lean/Expr/Basic.lean [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+end ConstantInfo
+
+namespace Expr
+
+/-! ### Declarations about `Expr` -/
+
+/-- If the expression is a constant, return that name. Otherwise return `Name.anonymous`. -/
+def constName (e : Expr) : Name :=
+e.constName?.getD Name.anonymous
```
Comment (gebner) [inline Mathlib/Lean/Expr/Basic.lean:101]:
```suggestion
  e.constName?.getD Name.anonymous
```

File: Mathlib/Lean/Expr/ReplaceRec.lean [hunk 3]
```diff
@@ -30,6 +30,22 @@ def replaceRec (f? : (Expr → Expr) → Expr → Option Expr) : Expr → Expr :
...
+/-- replaceRec under a monad. -/
+def replaceRecM [Monad M] (f? : (Expr → M Expr) → Expr → M (Option Expr)) : Expr → M Expr :=
+  fix fun r e => do
+    match ← f? r e with
+    | some x => return x
+    | none => traverseChildren r e
+
+/-- replaceRec except that bound variables are instantiated with free variables.
+This means that MetaM tactics can be used inside the replacement function. -/
+def replaceRecMeta [Monad M] [MonadLiftT MetaM M] [MonadControlT MetaM M]
```
Comment (gebner) [inline Mathlib/Lean/Expr/ReplaceRec.lean:42]: You might want to look at `Lean.Meta.transform`.

**Review by gebner — COMMENTED** [review]
LGTM
bors d+

> implement a version of replaceRecM that correctly instantiates free vars.

I'd try to use `Lean.Meta.transform` instead, both for consistency and caching.

File: Mathlib/Lean/Expr/ReplaceRec.lean (hunk 3 above)
Comment (EdAyers) [reply Mathlib/Lean/Expr/ReplaceRec.lean:42]: Yeah Lean.Meta.Transform doesn't let you recursively call the transform function which led to issues when porting to_additive

**Pull request closed without merging**